### PR TITLE
[compiled autograd] use same graph node names as AOTDispatcher

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133541
* __->__ #133148
* #133115

FIXES https://github.com/pytorch/pytorch/issues/132939


Compiled autograd's trace of the AOT backward may result in some additional ops e.g. clone to make contiguous, trace_wrapped HOPs, so the graphs may be slightly offset from each other


hf_Whisper example: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpNv89Pu/index.html
fsdp2 example: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpPdKssS/rank_0/index.html
Unit test example: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpvoQsnl/index.html
```python
 ===== Compiled autograd graph =====                                                                                                                                                          
 <eval_with_key>.14 class CompiledAutograd(torch.nn.Module):                                                                                                                                  
    def forward(self, inputs, sizes, scalars, hooks):                                                                                                                                         
        # No stacktrace found for following nodes                                                                                                                                             
        getitem: "f32[]cpu" = inputs[0]                                                                                                                                                       
        aot1_primals_1: "f32[4]cpu" = inputs[1]                                                                                                                                               
        aot1_primals_2: "f32[4]cpu" = inputs[2]                                                                                                                                               
        aot0_sin: "f32[4]cpu" = inputs[3]                                                                                                                                                     
        aot0_cos: "f32[4]cpu" = inputs[4]                                                                                                                                                     
        getitem_5: "f32[4]cpu" = inputs[5];  inputs = None                                                                                                                                    
                                                                                                                                                                                              
         # File: /data/users/xmfan/a/pytorch/torch/_dynamo/compiled_autograd.py:444 in set_node_origin, code: SumBackward0 (NodeCall 1)                                                       
        expand: "f32[4]cpu" = torch.ops.aten.expand.default(getitem, [4]);  getitem = None                                                                                                    
                                                                                                                                                                                              
         # File: /data/users/xmfan/a/pytorch/torch/_dynamo/compiled_autograd.py:444 in set_node_origin, code: CompiledFunctionBackward1 (NodeCall 2)                                          
        aot1_tangents_1: "f32[4]cpu" = torch.ops.aten.clone.default(expand, memory_format = torch.contiguous_format);  expand = None                                                          
        aot1_sin_1: "f32[4]cpu" = torch.ops.aten.sin.default(aot1_primals_2);  aot1_primals_2 = None                                                                                          
        aot1_neg: "f32[4]cpu" = torch.ops.aten.neg.default(aot1_sin_1);  aot1_sin_1 = None                                                                                                    
        aot0_tangents_2: "f32[4]cpu" = torch.ops.aten.mul.Tensor(aot1_tangents_1, aot1_neg);  aot1_neg = None                                                                                 
        aot1_cos_1: "f32[4]cpu" = torch.ops.aten.cos.default(aot1_primals_1);  aot1_primals_1 = None                                                                                          
        aot0_tangents_1: "f32[4]cpu" = torch.ops.aten.mul.Tensor(aot1_tangents_1, aot1_cos_1);  aot1_tangents_1 = aot1_cos_1 = None                                                           
                                                                                                                                                                                              
         # File: /data/users/xmfan/a/pytorch/torch/_dynamo/compiled_autograd.py:444 in set_node_origin, code: CompiledFunctionBackward0 (NodeCall 3)                                          
        aot0_neg: "f32[4]cpu" = torch.ops.aten.neg.default(aot0_sin);  aot0_sin = None                                                                                                        
        aot0_mul: "f32[4]cpu" = torch.ops.aten.mul.Tensor(aot0_tangents_2, aot0_neg);  aot0_tangents_2 = aot0_neg = None                                                                      
        aot0_mul_1: "f32[4]cpu" = torch.ops.aten.mul.Tensor(aot0_tangents_1, aot0_cos);  aot0_tangents_1 = aot0_cos = None                                                                    
        aot0_add: "f32[4]cpu" = torch.ops.aten.add.Tensor(aot0_mul, aot0_mul_1);  aot0_mul = aot0_mul_1 = None                                                                                
                                                                                                                                                                                              
         # File: /data/users/xmfan/a/pytorch/torch/_dynamo/compiled_autograd.py:444 in set_node_origin, code: torch::autograd::AccumulateGrad (NodeCall 4)                                    
        accumulate_grad_ = torch.ops.inductor.accumulate_grad_.default(getitem_5, aot0_add);  getitem_5 = aot0_add = accumulate_grad_ = None                                                  
        _exec_final_callbacks_stub = torch__dynamo_external_utils__exec_final_callbacks_stub();  _exec_final_callbacks_stub = None                                                            
        return []   
```

where aot1 is
```python
class GraphModule(torch.nn.Module):
    def forward(self, primals_1: "f32[4][1]cpu", primals_2: "f32[4][1]cpu", tangents_1: "f32[4][1]cpu"):
         # File: /data/users/xmfan/a/pytorch/test/inductor/test_compiled_autograd.py:2233 in torch_dynamo_resume_in_f_at_2232, code: return tmp1.sin() + tmp2.cos()
        sin_1: "f32[4][1]cpu" = torch.ops.aten.sin.default(primals_2);  primals_2 = None
        neg: "f32[4][1]cpu" = torch.ops.aten.neg.default(sin_1);  sin_1 = None
        mul: "f32[4][1]cpu" = torch.ops.aten.mul.Tensor(tangents_1, neg);  neg = None
        cos_1: "f32[4][1]cpu" = torch.ops.aten.cos.default(primals_1);  primals_1 = None
        mul_1: "f32[4][1]cpu" = torch.ops.aten.mul.Tensor(tangents_1, cos_1);  tangents_1 = cos_1 = None
        return (mul_1, mul)
```

and aot0 is
```python
class GraphModule(torch.nn.Module):
    def forward(self, sin: "f32[4][1]cpu", cos: "f32[4][1]cpu", tangents_1: "f32[4][1]cpu", tangents_2: "f32[4][1]cpu"):
         # File: /data/users/xmfan/a/pytorch/test/inductor/test_compiled_autograd.py:2231 in f, code: tmp2 = x.cos()
        neg: "f32[4][1]cpu" = torch.ops.aten.neg.default(sin);  sin = None
        mul: "f32[4][1]cpu" = torch.ops.aten.mul.Tensor(tangents_2, neg);  tangents_2 = neg = None
        
         # File: /data/users/xmfan/a/pytorch/test/inductor/test_compiled_autograd.py:2230 in f, code: tmp1 = x.sin()
        mul_1: "f32[4][1]cpu" = torch.ops.aten.mul.Tensor(tangents_1, cos);  tangents_1 = cos = None
        
         # File: /data/users/xmfan/a/pytorch/test/inductor/test_compiled_autograd.py:2230 in f, code: tmp1 = x.sin()
        add: "f32[4][1]cpu" = torch.ops.aten.add.Tensor(mul, mul_1);  mul = mul_1 = None
        return (add,)
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/inductor/test_compiled_autograd.py [hunk 1]
```diff
@@ -2102,6 +2102,33 @@ def forward(model, x):
...
+                torch._dynamo.graph_break()
+                return tmp1.sin() + tmp2.cos()
+
+            x = torch.randn(4, requires_grad=True)
+            out = f(x)
+            out.sum().backward()
+            yield x.grad
+
+        logs, ctx = logs_to_string(
+            torch._dynamo.compiled_autograd.__name__, "compiled_autograd_verbose"
```
Comment (yf225) [inline test/inductor/test_compiled_autograd.py:2122]: Curious does it only rename under `compiled_autograd_verbose`? (wonder if any downside to always rename regardless of verbose or not)

Comment (xmfan) [reply test/inductor/test_compiled_autograd.py:2241]: the concern is generally overhead, but in this case it is not on the hot path, so i can move it in a followup refactor

File: torch/_dynamo/compiled_autograd.py [hunk 2]
```diff
@@ -336,6 +339,82 @@ def runtime_wrapper(compiled_fn, inputs, sizes, scalars, hooks):
...
+        """
+        Renames nodes as they appear in the AOTDispatcher backward graphs, prefixed by AOT id
+        e.g. AOTDispatcher backward graph X's `sin_Y` -> `aotX_sin_Y`
+        """
+        if self.aot_graph_cls_name is None:
+            return
+
+        def is_similar(a: torch.fx.node.Node, b: torch.fx.node.Node):
+            if callable(a.target) and callable(b.target):
+                target_match = a.target.__qualname__ == b.target.__qualname__
```
Comment (jansel) [inline torch/_dynamo/compiled_autograd.py:352]: are all callables in python guaranteed to have a `__qualname__`?  (include C extension functions?)

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/compiled_autograd.py (hunk 2 above)
Comment (xmfan) [reply torch/_dynamo/compiled_autograd.py:352]: any functions yes, but for instances of classes that defined __call__, we have to set them manually. this actually breaks for HigherOrderOperator which doesn't set  __qualname__, i'll move the fix into this PR

Comment (jansel) [reply torch/_dynamo/compiled_autograd.py:352]: Maybe `getattr(a.target, "__qualname__", "<unset>")`

Comment (xmfan) [reply torch/_dynamo/compiled_autograd.py]: gonna use __name__ instead since we still want to rename HOPs

**Pull request closed without merging**